DRILL-6798: Planner changes to support semi-join.

HanumathRao opened the pull request:
This PR contains the planner changes for semi join optimization. It enables the Calcite's SemiJoinRule to transform the Join -- Agg combination to DrillSemiJoin. 
During the physical planning DrillSemiJoin is transformed to HashJoinPrel which can use SemiJoin information to build hash table efficiently.

For now the planner changes are disabled by default.  DRILL-6735 fixes the hash join changes to use this flag to efficiently build the hash table. The PR for DRILL-6735 will enable the semijoin optimization by default.

@amansinha100  Can you please help review this PR.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillSemiJoinRel.java:
```diff
@@ -0,0 +1,100 @@
+                                 RelNode left, RelNode right, JoinRelType joinType, boolean semiJoinDone) {
+    assert joinType == JoinRelType.INNER;
+    final JoinInfo joinInfo = JoinInfo.of(left, right, condition);
+    assert joinInfo.isEqui();
```
Asserts are only enabled in debug builds..best to use `Preconditions` checks.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillSemiJoinRel.java:
```diff
@@ -0,0 +1,100 @@
+
+  public DrillSemiJoinRel(
+          RelOptCluster cluster,
+          RelTraitSet traitSet,
```
Does the caller ensure that the `traitSet` for a SemiJoin only has the collation and distribution properties from the left side of the join ?

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/RuleInstance.java:
```diff
@@ -17,7 +17,9 @@
  */
 package org.apache.drill.exec.planner;
 
+import com.google.common.base.Preconditions;
```
Need to add the "shaded guava" prefix: `org.apache.drill.shaded.guava.`

HanumathRao commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/join/TestSemiJoin.java:
```diff
@@ -0,0 +1,104 @@
+import org.junit.experimental.categories.Category;
+
+@Category({SlowTest.class, OperatorTest.class})
+public class TestSemiJoin extends BaseTestQuery {
```
@amansinha100  Boaz tried out cases with in subqueries and normal joins.

HanumathRao commented:
@amansinha100  Thank you for the review. I have addressed the review comments.

@Ben-Zvi  Thank you for the comment. I have made changes accordingly.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillSemiJoinRel.java:
```diff
@@ -0,0 +1,130 @@
+          condition,
+          leftKeys,
+          rightKeys);
+    Preconditions.checkArgument(validateTraits(traitSet, left, right));
```
The intent of my previous question was to see if this traits validation is already happening in Calcite since this functionality ideally belongs there.   I suppose you checked that ?

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillSemiJoinRel.java:
```diff
@@ -0,0 +1,130 @@
+      if (trait.getTraitDef().getTraitClass().equals(RelCollation.class)) {
+        RelCollation collationTrait = (RelCollation)trait;
+        for (RelFieldCollation field : collationTrait.getFieldCollations()) {
+          if (bitSet.indexOf(field.getFieldIndex()) < 0) {
```
Should this be ` > 0` ?  since we want to check presence of the right side's field in the collation traits.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/HashJoinPrel.java:
```diff
@@ -50,14 +51,21 @@
-    super(cluster, traits, left, right, condition, joinType);
+      boolean isRowKeyJoin, int joinControl, boolean semiJoin) throws InvalidRelException {
+    super(cluster, traits, left, right, condition, joinType, semiJoin);
+    Preconditions.checkArgument(isSemiJoin && !swapped || swapped && !isSemiJoin || (!swapped && !isSemiJoin));
```
Why not simplify to :  
`if (isSemiJoin) {`
   `Preconditions.checkArgument(!swapped : 'swapping of inputs is not allowed for semi-joins') `
`}`

HanumathRao commented:
@amansinha100  Thank you for the review.
